lib/curl_libproxy.c

```diff
@@ -0,0 +1,69 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) 2022-2023, Jan-Michael Brummer, <jan-michael.brummer1@volkswagen.de>.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at http://curl.haxx.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#ifndef CURL_DISABLE_LIBPROXY
+
+#include "urldata.h"
+
+#include <proxy.h>
+/* The last 3 #include files should be in this order */
+#include "curl_printf.h"
+#include "curl_memory.h"
+#include "memdebug.h"
+
+void Curl_libproxy_cleanup(pxProxyFactory *proxy_factory)
+{
+  if (proxy_factory)
+    px_proxy_factory_free(proxy_factory);
+}
+
+char *Curl_libproxy_detect_proxy(struct Curl_easy *data, const char *url)
+{
+  char *result = NULL;
+
+  /* Initialization of libproxy is done here, as otherwise libproxy and curl
+   * will have a cycle issue
+   */
+  if (!data->proxy_factory)
+    data->proxy_factory = px_proxy_factory_new();
+
+  if (data->proxy_factory) {
+    char **libproxy_results = px_proxy_factory_get_proxies(data->proxy_factory, url);
```

Comment:
This function call is blocking and can potentially take seconds to complete, right? It seems very bad and can potentially disrupt a lot of other already ongoing transfers if that happens. This very basic API is simple, but I would call it **too simple** because of this.  We want it asynchronous to allow the rest of the library to continue while this single transfer figures out its proxy.

Also, this makes the libproxy library parse the provided URL (doesn't it?) and we know that having different URL parsers act on the same URL is a recipe for disaster and security problems. Will it use libcurl's URL parser?

Reply:
libproxy is only making use of libcurl for it's download part and the download time is limited to 30sec (we can reduce this if required). Furthermore it is setting NOPROXY to ensure we are not cycling.

Reply:
You are referring to your _multi_xx functions, right?

Reply:
I can provide you a threaded version with a callback but i'm currently not sure how to connect it with curl in a sane way. Maybe you have an advice here?

Reply:
If threading is the only way to avoid blocking, then I believe we should keep the libproxy use out of libcurl. libcurl is meant to be able to run *thousands* of parallel transfers in a single thread. Adding threads for this purpose complicates matters a lot and makes libcurl use (much) more memory. Asynchronous non-blocking internal behaviors should be done without threads to the furthest extent possible (the threaded name resolving is the only exception).

Reply:
What do you have in mind as an alternative? Polling for feedback? Actually we are free to implement the solution that suits curl.

Reply:
Yes, polling. Other non-blocking APIs typically provide a file descriptor that can be used to wait for activity on (if it does not complete immediately). Once there is activity on the descriptor, libcurl calls the necessary function, which does as much as possible without blocking. Over and over until the action is complete.

Reply:
Performance Test
----------------------

I thought it would be good to add some performance data to the discussion. Libproxy actually does the following steps to
identify the necessary proxy configuration:

- Lookup of it's configuration
- Downloading & parsing PAC files if necessary

For lookup I've done a n-runs for receiving the configuration using the different plugins on Linux. Result is:

KDE         : 0.000015 s
GNOME  : 0.000034 s
Env          : 0.000004 s
sysconfig: 0.000004 s

For **manual proxy configuration** we would be out right here. Next parts are for **automatic proxy** configuration:

Parsing of a larger PAC file (20k): 0.001124 s

The only variable time would be in case wpad or pac proxy servers are set in the configuration. In this case downloading the
PAC file is depending on libcurl to download it (with a hard limit of 30s). 

In my case using the company proxy i did have a this result:

Download: 0.001235 s

Please note that this file is downloaded once and then reused for the next calls. So in the worst initial case i do have measured
this value:

**0,002396 s**

And for the continuing runs:

**0.000966 s**

Reply:
We need to design this feature (and others) to work proper in the **worst cases**. Like when it runs on on anciently slow hardware, the name resolve takes multiple seconds, the loading of the PAC is done over a slow VPN across the globe and the file is hundreds of kilobytes of complicated JavaScript.

Reply:
We've been working on a user-mode proxy library that has threading support here: https://github.com/nmoinvaz/proxyres.

Reply:
@nmoinvaz I don't know what "threading support" means, but if you read my comments again here you see that I much prefer *not* using threads. Good old non-blocking API using the single thread.

Reply:
I only meant that the library pushes the proxy resolution to a thread pool if the specific OS does not have asynchronous proxy resolution support. As far as the library API goes, it does not block.

Reply:
I would not be against support *multiple* libraries for detecting proxy. Then the user can decide at build-time which one to use. We do this already for TLS, SSH, QUIC and more.